corehq/apps/cloudcare/utils.py

```diff
@@ -6,10 +6,70 @@
 from six.moves.urllib.parse import quote
 
 from corehq import toggles
-from corehq.apps.app_manager.dbaccessors import get_apps_in_domain
+from corehq.apps.app_manager.dbaccessors import (
+    get_app_ids_in_domain,
+    get_apps_in_domain,
+    get_latest_build_doc,
+    get_latest_build_id,
+    get_latest_released_app_doc,
+    get_latest_released_build_id,
+)
 from corehq.util.quickcache import quickcache
 
 
+def can_user_access_web_app(user, app):
+    """
+    :param user: either a WebUser or CommCareUser
+    :param app: app doc (not wrapped Application)
+    """
+    domain = app.get("domain")
+    # Backwards-compatibility - mobile users haven't historically required this permission
+    has_access_via_permission = user.is_commcare_user()
+    if user.is_web_user() or user.can_access_any_web_apps(domain):
+        has_access_via_permission = user.can_access_web_app(domain, app.get('copy_of', app.get('_id')))
+
+    has_access_via_group = True  # permission takes precedence over groups, so default to True
+    if toggles.WEB_APPS_PERMISSIONS_VIA_GROUPS.enabled(domain):
+        from corehq.apps.cloudcare.dbaccessors import (
+            get_application_access_for_domain,
+        )
+        app_access = get_application_access_for_domain(domain)
+        has_access_via_group = app_access.user_can_access_app(user, app)
+
+    return has_access_via_permission and has_access_via_group
+
+
+def get_latest_build_for_web_apps(domain, username, app_id):
+    if (toggles.CLOUDCARE_LATEST_BUILD.enabled(domain) or toggles.CLOUDCARE_LATEST_BUILD.enabled(username)):
+        return get_latest_build_doc(domain, app_id)
+    else:
+        return get_latest_released_app_doc(domain, app_id)
+
+
+def get_latest_build_id_for_web_apps(domain, username, app_id):
+    if (toggles.CLOUDCARE_LATEST_BUILD.enabled(domain) or toggles.CLOUDCARE_LATEST_BUILD.enabled(username)):
+        return get_latest_build_id(domain, app_id)
+    else:
+        return get_latest_released_build_id(domain, app_id)
+
+
+def get_web_apps_available_to_user(domain, user, is_preview=False, fetch_app_fn=get_latest_build_for_web_apps):
+    """
+    The fetch_app_fn is a function to fetch app docs, and should accept a domain, username and app_id if overridden
+    """
+    def is_web_app(app):
+        return app.get('cloudcare_enabled') or is_preview
+
+    apps = []
+    app_ids = get_app_ids_in_domain(domain)
+    for app_id in app_ids:
+        app = fetch_app_fn(domain, user.username, app_id)
+        if app and is_web_app(app) and can_user_access_web_app(user, app):
```

Comment:
Could you also call this new function in `FormplayerPreviewSingleApp.get`? It currently has some of the logic that you've so nicely encapsulated.

Reply:
Done in 4a68aef15fcd1818f42fe3a0fb1d27ac567bf644